experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -71,6 +83,187 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+    return failure();
+
+  // Create visited set
+  llvm::DenseSet<Operation *> visited;
+  visited.insert(specOp.getDataOut().getDefiningOp());
+
+  // Start traversal at the speculator output
+  for (Operation *succOp : specOp.getDataOut().getUsers()) {
+    routeBranchControlTraversal(visited, commitCtrl, succOp);
+  }
+
+  return success();
+}
+
+static LogicalResult findControlBranch(ModuleOp modOp, unsigned bb,
+                                       ConditionalBranchOp &cbranchOp) {
+  handshake::FuncOp funcOp = *modOp.getOps<handshake::FuncOp>().begin();
+  assert(funcOp && "funcOp not found");
+  auto handshakeBlocks = getLogicBBs(funcOp);
+
+  handshake::ConditionalBranchOp controlBranch;
+
+  // Iterate all Ops in the BB where we want to have Speculation
+  for (auto blockOp : handshakeBlocks.blocks.lookup(bb)) {
+    if ((controlBranch = dyn_cast<handshake::ConditionalBranchOp>(blockOp))) {
+      // SaveCommits should only be placed in BBs where there is a Backedge
+      // If this function is called, a backedge should exist
+      for (Value result : blockOp->getResults()) {
+        for (Operation *user : result.getUsers()) {
+          if (isBackedge(result, user)) {
+            cbranchOp = controlBranch;
+            return success();
+          }
+        }
+      }
+    }
+  }
+  llvm::errs() << "Could not find backedge within the Speculator's BB\n";
+  return failure();
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceSaveCommits() {
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+
+  // Don't do anything if there are no SaveCommits to place
+  if (placements.getPlacements<handshake::SpecSaveCommitOp>().size() == 0)
+    return success();
+
+  // The save commits are a result of a control branch being in the BB
+  // The control path for the SC needs to replicate the branch
+  unsigned bb = getLogicBB(specOp.getDataIn().getDefiningOp()).value();
+  ConditionalBranchOp controlBranch;
+  if (failed(findControlBranch(getOperation(), bb, controlBranch)))
+    return failure();
+
+  // To connect a Save-Commit, two control signals are sent from the Speculator
+  // and are merged before reaching the Save-Commit.
+  // The tokens take differents paths. One needs to always reach the SC,
+  // the other should follow the actual branches similarly to the Commits
+  builder.setInsertionPointAfterValue(specOp.getSCSaveCtrl());
+  auto branchDisc1 = builder.create<handshake::SpeculatingBranchOp>(
+      controlBranch.getLoc(),
+      controlBranch.getTrueResult(),      // specTagOperand
+      controlBranch.getConditionOperand() // Ctrl signal fromthe branch
+  );
+
+  // This branch will propagate the signal SCControl1 according to thecontrol
+  // branch condition, which comes from branchDisc1
+  auto branchDisc2 = builder.create<handshake::ConditionalBranchOp>(
+      branchDisc1.getLoc(), branchDisc1.getTrueResult(),
+      specOp.getSCSaveCtrl());
```

Comment:
this seems wrong to me, the save-like control signal shouldn't go through any branches? 
It tells the save-commits whether the last value they saw should be resent or dropped, it's not dependent on the control token?
Looking at figure 23 for this.

Reply:
This was wrong and has been changed. SCSaveCtrl -> SCCommitCtrl